### PR TITLE
Refactoring for layouts + loading indicator

### LINKED ISSUE #182: Add a default loading indicator between page in default boilerplate and docs
Something like [topbar](http://buunguyen.github.io/topbar/)?
A simple spinner?
A content placeholder?
topbar + a content placeholder?

### PR DESCRIPTION
**Pull request by MoOx**
Sorry for the big PR, maybe try to review one commit at a time :)

The refactoring supports current method, so no breaking changes. Show a notice to users.

Loading page looks like this:

screen shot 2016-04-11 at 00 30 27

The spinner in the middle only show if the delay is > 250ms, which avoid flash spinner (which sucks).
Note the top bar (logarithmic effect: start quick, and the more it loads, the less its fast). Better UX imo compared to most loading topbar which quickly load around 20% then go right to the 100%.

Boilerplate 404 is now like docs

screen shot 2016-04-11 at 00 30 24

I will add more components and layouts in default boilerplate after that (have local wip, but I found a bug and want to write test for it).

Will closes #182 and #357.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,47 @@
...
+  import StatinamicPageContainer from "statinamic/lib/PageContainer"
+
+  import Page from "../layouts/Page"
+  import PageError from "../layouts/PageError"
+  import PageLoading from "../layouts/PageLoading"
+  import Homepage from "../layouts/Homepage"
+
+  class PageContainer extends Component {
+    render() {
+      const { props } = this
```
Comment (thangngoc89) [inline CHANGELOG.md]: This is super confusing. Why not just `{ ...this.props }` ?

File: src/ContextProvider/index.js [hunk 2]
```diff
@@ -5,28 +5,28 @@ import { Component, PropTypes, Children } from "react"
 export default class StatinamicContext extends Component {
   props: {
     collection: StatinamicCollection,
-    layouts: Object,
+    layouts: Object, // deprecated
```
Comment (thangngoc89) [inline src/ContextProvider/index.js]: I know this is not the scope of this PR but flowtype props and propTypes is not the same here.

collection, layouts and metadata is required in flowtype.

**Comment by thangngoc89** [comment]
Awesome UX with topbar and loading spinner

File: src/ContextProvider/index.js (hunk 2 above)
Comment (MoOx) [inline src/ContextProvider/index.js]: You are right. But without collection or layout, we will have issue. Will check this before removing proptypes

File: CHANGELOG.md (hunk 1 above)
Comment (MoOx) [inline CHANGELOG.md]: I like to prepare myself for functional component (stateless or [stateful](https://github.com/reactjs/react-future/blob/ba5a275d08451756137559714d79f15055b9a14e/07%20-%20Returning%20State/01%20-%20Stateful%20Functions.js)) :D

Comment (MoOx) [inline CHANGELOG.md]: So now I like to extract props and state has const :)

Comment (thangngoc89) [inline CHANGELOG.md]: Thanks for useful link :)

**Pull request merged**